### PR TITLE
Fix at::optional compile problems on Windows CUDA.

### LINKED ISSUE #10329: at::optional does not work on nvcc Windows (CUDA 9.1)
Upstream issue at https://github.com/akrzemi1/Optional/issues/58

Possible resolutions:

- I've tested that a trivial program using https://github.com/martinmoene/optional-lite DOES compile successfully on Windows. Maybe we can switch to it.
- @ngimel suggested maybe CUDA 9.2 fixes the problem. I haven't been able to test this yet.

### PR DESCRIPTION
**Pull request by gchanan**
Fixes https://github.com/pytorch/pytorch/issues/10329.

### CODE REVIEW

File: aten/src/ATen/core/optional.h [hunk 1]
```diff
@@ -301,6 +301,14 @@ using OptionalBase = typename std::conditional<
 
 template <class T>
 class optional : private OptionalBase<T> {
+
+  template <class U>  // re-declaration for nvcc on Windows.
```
Comment (ezyang) [inline aten/src/ATen/core/optional.h]: I wish we knew what actual MSVC/NVCC bug we were triggering here, but it's not very obvious.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Can you cite #10329 in the commit message?

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/core/optional.h (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/core/optional.h]: Will do!

**Comment by gchanan** [comment]
I'll cite https://github.com/pytorch/pytorch/issues/10329 in the commit message.

**Comment by goldsborough** [comment]
wooooooooh 🎉

File: aten/src/ATen/core/optional.h (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/core/optional.h]: From messing around on the windows machine it's something like conditional template specifications aren't visible inside a class that extends a conditional template specification.  When I defined TensorBase as either of the conditional arguments it compiled fine, but as soon as the conditional was introduced it broke (even if the if was just true or false), even when I defined my own conditional.

**Pull request closed without merging**